File: tests/store/review.test.ts
import { expect, test, describe } from "bun:test";
import { AuthorStore } from "../../src/store/author";
import { BookStore } from "../../src/store/book";
import { ReviewStore } from "../../src/store/review";
import { newTestDB } from "../helpers";

describe("ReviewStore", () => {
  test("create", () => {
    const db = newTestDB();
    const as = new AuthorStore(db);
    const bs = new BookStore(db);
    const rs = new ReviewStore(db);
    const author = as.create("Author", "");
    const book = bs.create("Book", author.id, "", 2000);
    const r = rs.create(book.id, 5, "Excellent read.");
    expect(r.id).toBeGreaterThan(0);
    expect(r.rating).toBe(5);
    expect(r.book_id).toBe(book.id);
  });

  describe("listByBook", () => {
    function setup() {
      const db = newTestDB();
      const as = new AuthorStore(db);
      const bs = new BookStore(db);
      const rs = new ReviewStore(db);
      const author = as.create("Author", "");
      const book1 = bs.create("Book 1", author.id, "", 2000);
      const book2 = bs.create("Book 2", author.id, "", 2001);
      rs.create(book1.id, 5, "Great.");
      rs.create(book1.id, 3, "Average.");
      rs.create(book2.id, 4, "Good.");
      return { rs, book1, book2 };
    }

    test("book1 has two reviews", () => {
      const { rs, book1 } = setup();
      expect(rs.listByBook(book1.id).length).toBe(2);
    });
    test("book2 has one review", () => {
      const { rs, book2 } = setup();
      expect(rs.listByBook(book2.id).length).toBe(1);
    });
    test("unknown book returns empty slice", () => {
      const { rs } = setup();
      expect(rs.listByBook(99999).length).toBe(0);
    });
  });
});

File: internal/store/review_test.go
package store_test

import (
	"testing"

	"github.com/xebia/ai-assisted-se-claude-exercise/internal/store"
)

func TestReviewCreate(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)
	rs := store.NewReviewStore(db)

	author, _ := as.Create("Author", "")
	book, _ := bs.Create("Book", author.ID, "", 2000)

	r, err := rs.Create(book.ID, 5, "Excellent read.")
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if r.ID == 0 {
		t.Error("expected non-zero ID")
	}
	if r.Rating != 5 {
		t.Errorf("Rating = %d, want 5", r.Rating)
	}
	if r.BookID != book.ID {
		t.Errorf("BookID = %d, want %d", r.BookID, book.ID)
	}
}

func TestReviewListByBook(t *testing.T) {
	db := newTestDB(t)
	as := store.NewAuthorStore(db)
	bs := store.NewBookStore(db)
	rs := store.NewReviewStore(db)

	author, _ := as.Create("Author", "")
	book1, _ := bs.Create("Book 1", author.ID, "", 2000)
	book2, _ := bs.Create("Book 2", author.ID, "", 2001)

	rs.Create(book1.ID, 5, "Great.")
	rs.Create(book1.ID, 3, "Average.")
	rs.Create(book2.ID, 4, "Good.")

	t.Run("book1 has two reviews", func(t *testing.T) {
		reviews, err := rs.ListByBook(book1.ID)
		if err != nil {
			t.Fatalf("ListByBook: %v", err)
		}
		if len(reviews) != 2 {
			t.Errorf("len = %d, want 2", len(reviews))
		}
	})

	t.Run("book2 has one review", func(t *testing.T) {
		reviews, err := rs.ListByBook(book2.ID)
		if err != nil {
			t.Fatalf("ListByBook: %v", err)
		}
		if len(reviews) != 1 {
			t.Errorf("len = %d, want 1", len(reviews))
		}
	})

	t.Run("unknown book returns empty slice", func(t *testing.T) {
		reviews, err := rs.ListByBook(99999)
		if err != nil {
			t.Fatalf("ListByBook: %v", err)
		}
		if len(reviews) != 0 {
			t.Errorf("len = %d, want 0", len(reviews))
		}
	})
}
